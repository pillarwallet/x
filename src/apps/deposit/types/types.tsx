export type BalanceInfo = {
  type?: 'BalanceInfo';
  name: string;
  chain: string;
  address: string;
  decimals: number;
  symbol: string;
  logoURI: string;
  balance: string;
};

export type AddedAssets = {
  type?: 'AddedAsset';
  assetType: 'token' | 'nft';
  tokenId?: string;
  chain: string;
  tokenAddress: string;
  balance: number;
};

export type Network =
  | 'ethereum'
  | 'polygon'
  | 'gnosis'
  | 'base'
  | 'bnb smart chain'
  | 'optimism'
  | 'arbitrum';

export type TokenList = {
  address: string;
  chainId: number;
  decimals: string;
  name: string;
  symbol: string;
  logoURI: string;
};

export type BalancesByChain = {
  [chain: string]: BalanceInfo[];
};
