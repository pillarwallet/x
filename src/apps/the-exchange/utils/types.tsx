import { BridgingProvider } from '@etherspot/data-utils/dist/cjs/sdk/data/constants';
import { Route } from '@lifi/sdk';
import { Hex } from 'viem';

export enum CardPosition {
  SWAP = 'SWAP',
  RECEIVE = 'RECEIVE',
}

export type SwapType = {
  fromAmount: number;
  fromTokenAddress: string;
  fromChainId: number;
  fromTokenDecimals: number;
  toTokenAddress: string;
  toChainId: number;
  toTokenDecimals: number;
  slippage: number;
  fromAccountAddress?: string;
  provider?: BridgingProvider;
};

export type SwapOffer = {
  tokenAmountToReceive: number;
  offer: Route;
};

export type ChainType = {
  chainId: number;
  chainName: string;
};

export type StepTransaction = {
  to?: string;
  data?: Hex;
  value?: bigint;
  gasLimit?: bigint;
  gasPrice?: bigint;
  chainId?: number;
  type?: number | string;
  transactionType?: StepType;
};

export type StepType = 'swap' | 'cross' | 'lifi' | 'custom' | 'approval';
