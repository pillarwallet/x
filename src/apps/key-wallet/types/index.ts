export interface Asset {
  id: number;
  name: string;
  symbol: string;
  logo: string;
  balance: number;
  decimals: number;
  price: number;
  price_change_24h: number;
  contract: string;
  chainId: number;
  chainName: string;
  usdBalance: number;
}

export interface SendFormData {
  amount: string;
  recipient: string;
}

export interface TransactionStatus {
  hash: string;
  chainId: number;
  status: 'pending' | 'success' | 'failed';
  timestamp: number;
}

