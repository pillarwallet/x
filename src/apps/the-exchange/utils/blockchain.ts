import { BigNumber, BigNumberish } from 'ethers';
import { formatEther, formatUnits } from 'ethers/lib/utils';
import { decodeFunctionData, erc20Abi } from 'viem';

// types
import { Token } from '../../../services/tokensData';
import { StepTransaction } from './types';

export const processBigNumber = (val: BigNumber): number =>
  Number(val.toString());

export const processEth = (val: BigNumberish, dec: number): number => {
  if (typeof val === 'bigint') {
    return +parseFloat(formatEther(val)).toFixed(2);
  }

  return +parseFloat(formatUnits(val as BigNumberish, dec));
};

// Utility: get native token symbol for a chain
export const NATIVE_SYMBOLS: Record<number, string> = {
  1: 'ETH',
  100: 'xDAI',
  137: 'POL',
  10: 'ETH',
  42161: 'ETH',
  56: 'BNB',
  8453: 'ETH',
};

// Helper: Detect if a tx is a native fee step
export const isNativeFeeTx = (
  tx: StepTransaction,
  feeReceiver: string
): boolean => {
  return (
    typeof tx.to === 'string' &&
    typeof feeReceiver === 'string' &&
    tx.to.toLowerCase() === feeReceiver.toLowerCase()
  );
};

// Helper: Detect if a tx is an ERC20 (stablecoin or wrapped) fee step
export const isERC20FeeTx = (
  tx: StepTransaction,
  swapToken: Token
): boolean => {
  return (
    typeof tx.to === 'string' &&
    typeof swapToken.contract === 'string' &&
    tx.to.toLowerCase() === swapToken.contract.toLowerCase() &&
    tx.value === BigInt(0) &&
    typeof tx.data === 'string' &&
    tx.data.startsWith('0xa9059cbb')
  );
};

// Helper: Extract fee amount from tx
export const getFeeAmount = (
  tx: StepTransaction,
  swapToken: Token,
  decimals: number
): string => {
  if (tx.value && tx.data === '0x') {
    // Native
    return formatEther(tx.value);
  }
  if (isERC20FeeTx(tx, swapToken)) {
    try {
      const decoded = decodeFunctionData({
        abi: erc20Abi,
        data: tx.data || '0x',
      });
      if (decoded.args && typeof decoded.args[1] === 'bigint') {
        return formatUnits(decoded.args[1], decimals);
      }
    } catch (e) {
      return '0';
    }
  }
  return '0';
};

// Helper: Get fee symbol
export const getFeeSymbol = (
  tx: StepTransaction,
  swapToken: Token,
  chainId: number
): string => {
  if (tx.value && tx.data === '0x') {
    return NATIVE_SYMBOLS[chainId] || 'NATIVE';
  }
  return swapToken.symbol;
};
