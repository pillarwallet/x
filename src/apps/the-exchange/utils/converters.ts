export const hasThreeZerosAfterDecimal = (num: number): boolean => {
  const decimalPart = num.toString().split('.')[1] || '';
  return decimalPart.startsWith('000');
};

export const formatTokenAmount = (amount?: number) => {
  if (amount === undefined) return 0;
  return hasThreeZerosAfterDecimal(amount)
    ? amount.toFixed(8)
    : amount.toFixed(4);
};
