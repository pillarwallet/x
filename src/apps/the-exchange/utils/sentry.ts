import * as Sentry from '@sentry/react';
import { useWalletAddress } from '@etherspot/transaction-kit';

// Sentry configuration for the-exchange app
export const initSentryForExchange = () => {
  Sentry.setTag('app', 'the-exchange');
  Sentry.setTag('module', 'exchange');
};

// Utility to get fallback wallet address for logging
// This function should be called from within a React component context
// where the wallet address is available
export const fallbackWalletAddressForLogging = (): string => {
  // This is a utility function that should be called from within React components
  // The actual wallet address should be passed as a parameter or obtained via hook
  // For now, return unknown as this is a fallback utility
  return 'unknown_wallet_address';
};

// Enhanced Sentry logging with wallet address context
export const logExchangeEvent = (
  message: string,
  level: Sentry.SeverityLevel = 'info',
  extra?: Record<string, unknown>,
  tags?: Record<string, string>
) => {
  const walletAddress = fallbackWalletAddressForLogging();

  Sentry.withScope((scope) => {
    scope.setLevel(level);
    scope.setTag('wallet_address', walletAddress);
    scope.setTag('app_module', 'the-exchange');

    if (tags) {
      Object.entries(tags).forEach(([key, value]) => {
        scope.setTag(key, value);
      });
    }

    if (extra) {
      scope.setExtra('exchange_data', extra);
    }

    Sentry.captureMessage(message);
  });
};

// Log exchange errors with wallet address
export const logExchangeError = (
  error: Error | string,
  extra?: Record<string, unknown>,
  tags?: Record<string, string>
) => {
  const walletAddress = fallbackWalletAddressForLogging();

  Sentry.withScope((scope) => {
    scope.setLevel('error');
    scope.setTag('wallet_address', walletAddress);
    scope.setTag('app_module', 'the-exchange');
    scope.setTag('error_type', 'exchange_error');

    if (tags) {
      Object.entries(tags).forEach(([key, value]) => {
        scope.setTag(key, value);
      });
    }

    if (extra) {
      scope.setExtra('exchange_error_data', extra);
    }

    if (error instanceof Error) {
      Sentry.captureException(error);
    } else {
      Sentry.captureMessage(error, 'error');
    }
  });
};

// Generic operation logging function
export const logOperation = (
  operationType: string,
  operation: string,
  data: Record<string, unknown>,
  walletAddress?: string
) => {
  Sentry.withScope((scope) => {
    scope.setLevel('info');
    scope.setTag(
      'wallet_address',
      walletAddress || fallbackWalletAddressForLogging()
    );
    scope.setTag('app_module', 'the-exchange');
    scope.setTag('operation_type', operationType);
    scope.setTag(`${operationType}_operation`, operation);

    scope.setExtra(`${operationType}_data`, data);

    Sentry.captureMessage(
      `${operationType.charAt(0).toUpperCase() + operationType.slice(1)} operation: ${operation}`,
      'info'
    );
  });
};

// Log swap operations
export const logSwapOperation = (
  operation: string,
  data: Record<string, unknown>,
  walletAddress?: string
) => {
  logOperation('swap', operation, data, walletAddress);
};

// Log token operations
export const logTokenOperation = (
  operation: string,
  tokenData: Record<string, unknown>,
  walletAddress?: string
) => {
  logOperation('token', operation, tokenData, walletAddress);
};

// Log offer operations
export const logOfferOperation = (
  operation: string,
  offerData: Record<string, unknown>,
  walletAddress?: string
) => {
  logOperation('offer', operation, offerData, walletAddress);
};

// Log transaction operations
export const logTransactionOperation = (
  operation: string,
  transactionData: Record<string, unknown>,
  walletAddress?: string
) => {
  logOperation('transaction', operation, transactionData, walletAddress);
};

// Log user interactions
export const logUserInteraction = (
  interaction: string,
  data: Record<string, unknown>,
  walletAddress?: string
) => {
  Sentry.withScope((scope) => {
    scope.setLevel('info');
    scope.setTag(
      'wallet_address',
      walletAddress || fallbackWalletAddressForLogging()
    );
    scope.setTag('app_module', 'the-exchange');
    scope.setTag('interaction_type', 'user_action');
    scope.setTag('user_interaction', interaction);

    scope.setExtra('interaction_data', data);

    Sentry.captureMessage(`User interaction: ${interaction}`, 'info');
  });
};

// Log performance metrics
export const logPerformanceMetric = (
  metric: string,
  value: number,
  unit: string = 'ms',
  walletAddress?: string
) => {
  Sentry.withScope((scope) => {
    scope.setLevel('info');
    scope.setTag(
      'wallet_address',
      walletAddress || fallbackWalletAddressForLogging()
    );
    scope.setTag('app_module', 'the-exchange');
    scope.setTag('metric_type', 'performance');
    scope.setTag('metric_name', metric);

    scope.setExtra('performance_data', {
      metric,
      value,
      unit,
      timestamp: new Date().toISOString(),
    });

    Sentry.captureMessage(
      `Performance metric: ${metric} = ${value}${unit}`,
      'info'
    );
  });
};

// Hook to get wallet address for logging
export const useWalletAddressForLogging = () => {
  const walletAddress = useWalletAddress();
  return walletAddress || 'unknown_wallet_address';
};

// Note: Error boundary removed due to TypeScript/JSX compatibility issues
// Use Sentry's default error boundary or create a separate React component file

// Transaction monitoring for exchange operations
export const startExchangeTransaction = (
  operation: string,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _data: Record<string, unknown> = {},
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _walletAddress?: string
) => {
  return Sentry.startSpan(
    {
      name: `exchange.${operation}`,
      op: 'exchange.operation',
    },
    (span) => {
      // Note: Span API has changed in Sentry v10
      // Properties are set via the span context instead
      return span;
    }
  );
};

// Breadcrumb utilities for exchange
export const addExchangeBreadcrumb = (
  message: string,
  category: string = 'exchange',
  data?: Record<string, unknown>,
  level: Sentry.SeverityLevel = 'info'
) => {
  const walletAddress = fallbackWalletAddressForLogging();

  Sentry.addBreadcrumb({
    message,
    category,
    level,
    data: {
      ...data,
      wallet_address: walletAddress,
      app_module: 'the-exchange',
    },
  });
};
