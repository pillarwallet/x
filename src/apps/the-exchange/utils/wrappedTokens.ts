export const NATIVE_TOKEN_ADDRESSES = new Set([
  '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
  '0x0000000000000000000000000000000000000000',
]);

// Not including XDAI below
export const WRAPPED_NATIVE_TOKEN_ADDRESSES: Record<number, string> = {
  // Ethereum
  1: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2', // WETH
  // Polygon
  137: '0x0d500B1d8E8eF31E21C99d1Db9A6444d3ADf1270', // WMATIC
  // Optimism
  10: '0x4200000000000000000000000000000000000006', // WETH
  // Arbitrum
  42161: '0x82aF49447D8a07e3bd95BD0d56f35241523fBab1', // WETH
  // Base
  8453: '0x4200000000000000000000000000000000000006', // WETH
  // BNB
  56: '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c', // WBNB
};

export const isWrappedToken = (tokenAddress: string, chainId: number) => {
  return (
    tokenAddress.toLowerCase() ===
    WRAPPED_NATIVE_TOKEN_ADDRESSES[chainId]?.toLowerCase()
  );
};

export const isNativeToken = (address: string) =>
  NATIVE_TOKEN_ADDRESSES.has(address.toLowerCase());

export const getWrappedTokenAddressIfNative = (
  tokenAddress: string,
  chainId: number
): string => {
  if (isNativeToken(tokenAddress)) {
    const wrappedAddress = WRAPPED_NATIVE_TOKEN_ADDRESSES[chainId];
    // fallback to token address in the case of XDAI token
    return wrappedAddress || tokenAddress;
  }
  return tokenAddress;
};
