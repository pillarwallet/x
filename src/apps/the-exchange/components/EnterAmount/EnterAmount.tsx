import _ from 'lodash';
import { useEffect, useState } from 'react';
import useTransactionKit from '../../../../hooks/useTransactionKit';

// reducer
import {
  setAmountReceive,
  setAmountSwap,
  setBestOffer,
  setIsAboveLimit,
  setIsOfferLoading,
  setUsdPriceReceiveToken,
  setUsdPriceSwapToken,
} from '../../reducer/theExchangeSlice';

// services
import {
  Token,
  chainNameToChainIdTokensData,
} from '../../../../services/tokensData';

// hooks
import useOffer from '../../hooks/useOffer';
import { useAppDispatch, useAppSelector } from '../../hooks/useReducerHooks';

// utils
import { formatExponential } from '../../../../utils/number';
import {
  addExchangeBreadcrumb,
  logExchangeError,
  logUserInteraction,
} from '../../utils/sentry';

// types
import { CardPosition, SwapOffer } from '../../utils/types';

// components
import BodySmall from '../Typography/BodySmall';

// images
import ReceiveArrow from '../../images/receive-arrow.png';
import SendArrow from '../../images/send-arrow.png';
import ExchangeOffer from './ExchangeOffer';

type EnterAmountProps = {
  type: CardPosition;
  tokenSymbol?: string;
  tokenBalance?: number;
  tokenChain?: string;
  isDeploymentCostLoading?: boolean;
  deploymentCost?: number;
};

const EnterAmount = ({
  type,
  tokenSymbol,
  tokenBalance,
  tokenChain,
  isDeploymentCostLoading,
  deploymentCost,
}: EnterAmountProps) => {
  const dispatch = useAppDispatch();
  const { walletAddress } = useTransactionKit();
  const amountSwap = useAppSelector((state) => state.swap.amountSwap as number);
  const amountReceive = useAppSelector(
    (state) => state.swap.amountReceive as number
  );
  const usdPriceSwapToken = useAppSelector(
    (state) => state.swap.usdPriceSwapToken as number
  );
  const usdPriceReceiveToken = useAppSelector(
    (state) => state.swap.usdPriceReceiveToken as number
  );
  const swapToken = useAppSelector((state) => state.swap.swapToken as Token);
  const receiveToken = useAppSelector(
    (state) => state.swap.receiveToken as Token
  );
  const bestOffer = useAppSelector(
    (state) => state.swap.bestOffer as SwapOffer
  );
  const isOfferLoading = useAppSelector(
    (state) => state.swap.isOfferLoading as boolean
  );
  const isAboveLimit = useAppSelector(
    (state) => state.swap.isAboveLimit as boolean
  );

  const [inputValue, setInputValue] = useState<string>('');
  const { getBestOffer } = useOffer();
  const [isNoOffer, setIsNoOffer] = useState<boolean>(false);

  // get usd price only when swap token changes
  useEffect(() => {
    if (swapToken) {
      if (!swapToken.price) {
        console.error(
          `Failed to fetch USD price of token: ${swapToken.symbol} ${swapToken.name} on ${swapToken.blockchain}`
        );
        dispatch(setUsdPriceSwapToken(0));
      }

      if (swapToken.price) {
        dispatch(setUsdPriceSwapToken(swapToken.price));
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [swapToken]);

  // get usd price only when receive token changes
  useEffect(() => {
    if (receiveToken) {
      if (!receiveToken.price) {
        console.error(
          `Failed to fetch USD price of token: ${receiveToken.symbol} ${receiveToken.name} on ${receiveToken.blockchain}`
        );
        dispatch(setUsdPriceReceiveToken(0));
      }

      if (receiveToken.price) {
        dispatch(setUsdPriceReceiveToken(receiveToken.price));
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [receiveToken]);

  // Gets the best swap offer
  const getOffer = async () => {
    addExchangeBreadcrumb('Getting best swap offer', 'offer', {
      amountSwap,
      swapToken: swapToken?.symbol,
      receiveToken: receiveToken?.symbol,
      walletAddress,
    });

    const params = {
      fromAmount: amountSwap ?? 0,
      fromTokenAddress: swapToken?.contract ?? '',
      fromChainId: chainNameToChainIdTokensData(swapToken?.blockchain) ?? 0,
      fromTokenDecimals: swapToken?.decimals ?? 0,
      toTokenAddress: receiveToken?.contract ?? '',
      toChainId: chainNameToChainIdTokensData(receiveToken?.blockchain) ?? 0,
      toTokenDecimals: receiveToken?.decimals ?? 0,
      slippage: 0.03,
    };

    const offer = await getBestOffer(params).catch((e) => {
      logExchangeError(
        e,
        {
          operation: 'get_best_offer',
          params,
          walletAddress,
        },
        {
          component: 'EnterAmount',
          method: 'getOffer',
        }
      );

      console.error(
        'Sorry, an error occurred while trying to fetch the best swap offer. Please try again.',
        e
      );
      return {} as SwapOffer;
    });

    if (offer && Object.keys(offer as SwapOffer).length && receiveToken) {
      dispatch(setAmountReceive(offer?.tokenAmountToReceive));
      addExchangeBreadcrumb('Best offer received', 'offer', {
        offer: offer?.tokenAmountToReceive,
        walletAddress,
      });
    } else {
      setIsNoOffer(true);
      addExchangeBreadcrumb('No offer available', 'offer', {
        walletAddress,
      });
    }

    dispatch(setIsOfferLoading(false));

    return offer;
  };

  // Debounced getOffer function with 2-second delay
  const debouncedGetOffer = _.debounce(() => {
    getOffer().then((offer) => {
      if (offer) dispatch(setBestOffer(offer));
    });
  }, 2000);

  // Similar function being used for the SendModal and AccountModal to check tokens balances
  const tokenBalanceLimit = (tokenAmount: number, balance: number) => {
    if (!swapToken && !balance) return undefined;

    // Check if the value exceeds the max token amount limit
    if (tokenAmount > balance - (deploymentCost ?? 0)) {
      return `The maximum amount of ${swapToken?.symbol}${deploymentCost && deploymentCost > 0 ? ' spendable' : ''} in your wallet is ${formatExponential(balance - (deploymentCost ?? 0))} ${swapToken?.symbol}${deploymentCost && deploymentCost > 0 ? ` because your wallet is currently undeployed on ${tokenChain}` : ''} - please change the amount and try again`;
    }

    return undefined;
  };

  useEffect(() => {
    if (amountSwap) {
      setInputValue(amountSwap.toString());
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [swapToken, receiveToken]);

  // getOffer will be called every time the swap amount or the swap/receive token is changed
  useEffect(() => {
    dispatch(setBestOffer(undefined));
    if (amountSwap && swapToken && receiveToken && !isAboveLimit) {
      dispatch(setIsOfferLoading(true));
      debouncedGetOffer();
    }
    // Clean-up debounce on component unmount
    return () => {
      dispatch(setIsOfferLoading(false));
      debouncedGetOffer.cancel();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [amountSwap, swapToken, receiveToken]);

  // When the token amount value changes, the input value reflects that change
  const handleTokenAmountChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;
    setInputValue(value);

    logUserInteraction('token_amount_changed', {
      type,
      value,
      tokenSymbol,
      tokenBalance,
      deploymentCost,
      walletAddress,
    });

    if (type === CardPosition.SWAP && swapToken) {
      dispatch(setAmountSwap(Number(value)));
    }

    if (tokenBalance && Number(value) > tokenBalance - (deploymentCost ?? 0)) {
      dispatch(setIsAboveLimit(true));
      addExchangeBreadcrumb('Amount above limit', 'validation', {
        value,
        tokenBalance,
        deploymentCost,
        walletAddress,
      });
    }
    if (tokenBalance && Number(value) <= tokenBalance - (deploymentCost ?? 0)) {
      dispatch(setIsAboveLimit(false));
    }
  };

  return (
    <div id="enter-amount-exchange" className="flex flex-col gap-1 group">
      <BodySmall className="group-hover:text-black_grey/[.4] font-normal">
        {tokenSymbol}
      </BodySmall>
      {type === CardPosition.SWAP ? (
        <>
          <input
            id="input-enter-amount-exchange"
            type="number"
            step="any"
            value={inputValue}
            onChange={(e) => handleTokenAmountChange(e)}
            placeholder="0"
            style={{ width: '100%' }}
            className="text-black_grey font-normal !text-3xl outline-none focus:outline-none focus:ring-0 focus:bg-[#292D32]/[.05] focus:border-b focus:border-b-black_grey group-hover:bg-[#292D32]/[.05] group-hover:border-b group-hover:border-b-black_grey"
            data-testid="enter-amount-input"
          />
          {tokenBalanceLimit(Number(inputValue), tokenBalance || 0) &&
            !isDeploymentCostLoading && (
              <BodySmall
                id="token-balance-limit-exchange"
                data-testid="error-max-limit"
              >
                {tokenBalanceLimit(Number(inputValue), tokenBalance || 0)}
              </BodySmall>
            )}
        </>
      ) : (
        <ExchangeOffer
          isOfferLoading={isOfferLoading}
          isNoOffer={isNoOffer}
          bestOffer={bestOffer}
        />
      )}
      <div className="flex justify-between">
        <BodySmall
          id="usd-conversion-exchange"
          className="group-hover:text-black_grey/[.4]"
        >
          $
          {type === CardPosition.SWAP
            ? (usdPriceSwapToken * amountSwap).toFixed(2)
            : (usdPriceReceiveToken * amountReceive).toFixed(2)}
        </BodySmall>
        <img
          src={type === CardPosition.SWAP ? SendArrow : ReceiveArrow}
          alt={type === CardPosition.SWAP ? 'Send' : 'Receive'}
        />
      </div>
    </div>
  );
};

export default EnterAmount;
