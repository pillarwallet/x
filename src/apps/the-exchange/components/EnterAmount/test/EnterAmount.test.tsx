import { fireEvent, render, waitFor } from '@testing-library/react';
import renderer, { act } from 'react-test-renderer';

// provider
import { Provider } from 'react-redux';
import { store } from '../../../../../store';

// reducer
import {
  setAmountReceive,
  setAmountSwap,
  setBestOffer,
  setIsOfferLoading,
  setIsReceiveOpen,
  setIsSwapOpen,
  setReceiveChain,
  setReceiveToken,
  setReceiveTokenData,
  setSearchTokenResult,
  setSwapChain,
  setSwapToken,
  setSwapTokenData,
  setUsdPriceReceiveToken,
  setUsdPriceSwapToken,
} from '../../../reducer/theExchangeSlice';

// types
import { AccountBalancesListenerRef } from '../../../../../providers/AccountBalancesProvider';
import { CardPosition } from '../../../utils/types';

// components
import EnterAmount from '../EnterAmount';

const mockTokenAssets = [
  {
    address: '0x01',
    name: 'Ether',
    symbol: 'ETH',
    chainId: 1,
    decimals: 18,
    icon: 'iconEth.png',
  },
  {
    address: '0x02',
    name: 'POL',
    symbol: 'POL',
    chainId: 137,
    decimals: 18,
    icon: 'iconMatic.png',
  },
];

// Mock transaction-kit hooks being used
jest.mock('@etherspot/transaction-kit', () => ({
  __esModule: true,
  useEtherspotSwaps: jest.fn().mockReturnValue({
    getOffers: jest.fn().mockResolvedValue([]),
    prepareCrossChainOfferTransactions: jest.fn().mockResolvedValue({}),
    getQuotes: jest.fn().mockResolvedValue({}),
  }),
  useEtherspotPrices: jest.fn().mockReturnValue({
    getPrice: jest.fn().mockResolvedValue({ usd: 1200 }),
    getPrices: jest.fn(),
  }),
  useWalletAddress: jest.fn().mockReturnValue({
    walletAddress: jest.fn(),
  }),
  useEtherspotUtils: jest.fn().mockReturnValue({
    isZeroAddress: jest.fn(),
    addressesEqual: jest.fn(),
  }),
  useEtherspotBalances: jest.fn().mockReturnValue({
    getAccountBalances: jest.fn(),
  }),
}));

// Mock useAccountBalances hook
jest.mock('../../../../../hooks/useAccountBalances', () => ({
  __esModule: true,
  default: jest.fn().mockReturnValue({
    listenerRef: { current: {} as AccountBalancesListenerRef },
    data: {
      balances: {
        '0x01': { balance: '0.2', usdValue: '6000' },
        '0x02': { balance: '20', usdValue: '6000' },
      },
      updateData: false,
      setUpdateData: jest.fn(),
    },
  }),
}));

describe('<EnterAmount />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    act(() => {
      store.dispatch(setSwapTokenData(mockTokenAssets));
      store.dispatch(setReceiveTokenData(mockTokenAssets));
      store.dispatch(setIsSwapOpen(false));
      store.dispatch(setIsReceiveOpen(false));
      store.dispatch(setSwapChain({ chainId: 1, chainName: 'Ethereum' }));
      store.dispatch(setReceiveChain({ chainId: 137, chainName: 'Polygon' }));
      store.dispatch(setSwapToken(mockTokenAssets[0]));
      store.dispatch(setReceiveToken(mockTokenAssets[1]));
      store.dispatch(setAmountSwap(0.1));
      store.dispatch(setAmountReceive(10));
      store.dispatch(setBestOffer(undefined));
      store.dispatch(setSearchTokenResult([]));
      store.dispatch(setUsdPriceSwapToken(1200));
      store.dispatch(setUsdPriceReceiveToken(0.4));
      store.dispatch(setIsOfferLoading(false));
    });
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <EnterAmount
            type={CardPosition.SWAP}
            tokenSymbol={store.getState().swap.swapToken.symbol}
          />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('handles token amount change in Swap card', async () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <EnterAmount
          type={CardPosition.SWAP}
          tokenSymbol={store.getState().swap.swapToken.symboll}
        />
      </Provider>
    );

    act(() => {
      store.dispatch(setAmountSwap(0));
      store.dispatch(setUsdPriceSwapToken(0));
    });

    const inputElement = getByTestId('enter-amount-input');
    fireEvent.change(inputElement, { target: { value: '50' } });

    await waitFor(() => {
      expect(store.getState().swap.amountSwap).toEqual(50);
      expect(store.getState().swap.usdPriceSwapToken).toEqual(1200);
    });
  });
});
