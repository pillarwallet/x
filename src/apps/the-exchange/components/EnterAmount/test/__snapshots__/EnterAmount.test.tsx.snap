// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<EnterAmount /> > renders correctly and matches snapshot 1`] = `
<div
  className="flex flex-col gap-1 group"
  id="enter-amount-exchange"
>
  <p
    className="text-xs font-normal text-black_grey group-hover:text-black_grey/[.4] font-normal"
  />
  <input
    className="text-black_grey font-normal !text-3xl outline-none focus:outline-none focus:ring-0 focus:bg-[#292D32]/[.05] focus:border-b focus:border-b-black_grey group-hover:bg-[#292D32]/[.05] group-hover:border-b group-hover:border-b-black_grey"
    data-testid="enter-amount-input"
    id="input-enter-amount-exchange"
    onChange={[Function]}
    placeholder="0"
    step="any"
    style={
      {
        "width": "100%",
      }
    }
    type="number"
    value=""
  />
  <div
    className="flex justify-between"
  >
    <p
      className="text-xs font-normal text-black_grey group-hover:text-black_grey/[.4]"
    >
      $
      0.01
    </p>
    <img
      alt="Send"
      src="/src/apps/the-exchange/images/send-arrow.png"
    />
  </div>
</div>
`;
